## 数据采集和理解

In [36]:
#设置ast_node_interactivity = "all"使得可以同时输出多条语句
from IPython.core.interactiveshell import InteractiveShell
InteractiveShell.ast_node_interactivity = "all"
#导入包
import pandas as pd
import numpy as np
#导入数据
train=pd.read_csv(r'E:\python\data\titanic\train.csv')
test=pd.read_csv(r'E:\python\data\titanic\test.csv')
print('训练集数据规模:{}'.format(train.shape))
print('测试集数据规模:{}'.format(test.shape))

训练集数据规模:(891, 12)
测试集数据规模:(418, 11)


训练数据集比测试数据集的列多一个，即Survived值。由于它是预测的生存值，所以，在测试数据集中没有。

In [35]:
#查看训练集信息
train.head()
test.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

PassengerId  Pclass                                          Name     Sex  \
0          892       3                              Kelly, Mr. James    male   
1          893       3              Wilkes, Mrs. James (Ellen Needs)  female   
2          894       2                     Myles, Mr. Thomas Francis    male   
3          895       3                              Wirz, Mr. Albert    male   
4          896       3  Hirvonen, Mrs. Alexander (Helga E Lindqvist)  female   

    Age  SibSp  Parch   Ticket     Fare Cabin Embarked  
0  34.5      0      0   330911   7.8292   NaN        Q  
1  47.0      1      0   363272   7.0000   NaN        S  
2  62.0      0      0   240276   9.6875   NaN        Q  
3  27.0      0      0   315154   8.6625   NaN        S  
4  22.0      1      1  3101298  12.2875   NaN        S

为了方便对训练数据和测试数据进行清洗，将训练数据和测试数据进行合并

In [4]:
#通过设置ignore_index=True参数，合并后的数据集会重新生成一个index
full=pd.concat([train,test],ignore_index=True)
full.head()

Age Cabin Embarked     Fare  \
0  22.0   NaN        S   7.2500   
1  38.0   C85        C  71.2833   
2  26.0   NaN        S   7.9250   
3  35.0  C123        S  53.1000   
4  35.0   NaN        S   8.0500   

                                                Name  Parch  PassengerId  \
0                            Braund, Mr. Owen Harris      0            1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...      0            2   
2                             Heikkinen, Miss. Laina      0            3   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)      0            4   
4                           Allen, Mr. William Henry      0            5   

   Pclass     Sex  SibSp  Survived            Ticket  
0       3    male      1       0.0         A/5 21171  
1       1  female      1       1.0          PC 17599  
2       3  female      0       1.0  STON/O2. 3101282  
3       1  female      1       1.0            113803  
4       3    male      0       0.0            373450

针对每一个字段做一个简单的解释：
PassengerId: 乘客ID；

Survived: 生存情况，0代表不幸遇难，1代表存活；

Pclass: 仓位等级，1为一等舱，2为二等舱，3为三等舱；

Name: 乘客姓名；

Sex: 性别；

Age: 年龄；

SibSp: 乘客在船上的兄妹姐妹数/配偶数（即同代直系亲属数）；

Parch: 乘客在船上的父母数/子女数（即不同代直系亲属数）；

Ticket: 船票编号；

Fare: 船票价格；

Cabin: 客舱号；

Embarked: 登船港口（S: Southampton; C: Cherbourg Q: Queenstown）

In [5]:
#查看数据描述性统计
full.describe()

Age         Fare        Parch  PassengerId       Pclass  \
count  1046.000000  1308.000000  1309.000000  1309.000000  1309.000000   
mean     29.881138    33.295479     0.385027   655.000000     2.294882   
std      14.413493    51.758668     0.865560   378.020061     0.837836   
min       0.170000     0.000000     0.000000     1.000000     1.000000   
25%      21.000000     7.895800     0.000000   328.000000     2.000000   
50%      28.000000    14.454200     0.000000   655.000000     3.000000   
75%      39.000000    31.275000     0.000000   982.000000     3.000000   
max      80.000000   512.329200     9.000000  1309.000000     3.000000   

             SibSp    Survived  
count  1309.000000  891.000000  
mean      0.498854    0.383838  
std       1.041658    0.486592  
min       0.000000    0.000000  
25%       0.000000    0.000000  
50%       0.000000    0.000000  
75%       1.000000    1.000000  
max       8.000000    1.000000

In [6]:
#因为，describe()函数只能查看数据类型的描述统计信息，无法查看类似字符类型的信息。故，需用info()函数进一步查看每一列的数据信息。
full.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1309 entries, 0 to 1308
Data columns (total 12 columns):
Age            1046 non-null float64
Cabin          295 non-null object
Embarked       1307 non-null object
Fare           1308 non-null float64
Name           1309 non-null object
Parch          1309 non-null int64
PassengerId    1309 non-null int64
Pclass         1309 non-null int64
Sex            1309 non-null object
SibSp          1309 non-null int64
Survived       891 non-null float64
Ticket         1309 non-null object
dtypes: float64(3), int64(4), object(5)
memory usage: 122.8+ KB


数据的总行数为1309行，其中，Age一栏中263列有缺失项；Fare一栏中1列有缺失项；Survived一栏只有891列，刚好对应训练数据集的行数。除了Age和Fare以外，Cabin/Embarked也有缺失项。
也可以用另一个命令，查看缺失项信息

In [7]:
full.isnull().sum()

Age             263
Cabin          1014
Embarked          2
Fare              1
Name              0
Parch             0
PassengerId       0
Pclass            0
Sex               0
SibSp             0
Survived        418
Ticket            0
dtype: int64

## 数据清洗

如果是数值类型，使用平均值或者中位数进行填充

年龄(Age) 最小值为0.17，不存在0值，其数据缺失率为263/1309=20.09%，由于Age的平均数与中位数接近，故选择平均值作为缺失项的填充值。

In [8]:
full['Age']=full['Age'].fillna(full['Age'].mean())

船票价格(Fare)一栏数据缺失项仅为一行，且存在票价为0的记录，如下：

In [9]:
full.loc[full['Fare']==0,:]

Age        Cabin Embarked  Fare  \
179   36.000000          NaN        S   0.0   
263   40.000000          B94        S   0.0   
271   25.000000          NaN        S   0.0   
277   29.881138          NaN        S   0.0   
302   19.000000          NaN        S   0.0   
413   29.881138          NaN        S   0.0   
466   29.881138          NaN        S   0.0   
481   29.881138          NaN        S   0.0   
597   49.000000          NaN        S   0.0   
633   29.881138          NaN        S   0.0   
674   29.881138          NaN        S   0.0   
732   29.881138          NaN        S   0.0   
806   39.000000          A36        S   0.0   
815   29.881138         B102        S   0.0   
822   38.000000          NaN        S   0.0   
1157  29.881138          NaN        S   0.0   
1263  49.000000  B52 B54 B56        S   0.0   

                                       Name  Parch  PassengerId  Pclass   Sex  \
179                     Leonard, Mr. Lionel      0          180       3  male   
263                   Harrison, Mr. William      0          264       1  male   
271            Tornquist, Mr. William Henry      0          272       3  male   
277             Parkes, Mr. Francis "Frank"      0          278       2  male   
302         Johnson, Mr. William Cahoone Jr      0          303       3  male   
413          Cunningham, Mr. Alfred Fleming      0          414       2  male   
466                   Campbell, Mr. William      0          467       2  male   
481        Frost, Mr. Anthony Wood "Archie"      0          482       2  male   
597                     Johnson, Mr. Alfred      0          598       3  male   
633           Parr, Mr. William Henry Marsh      0          634       1  male   
674              Watson, Mr. Ennis Hastings      0          675       2  male   
732                    Knight, Mr. Robert J      0          733       2  male   
806                  Andrews, Mr. Thomas Jr      0          807       1  male   
815                        Fry, Mr. Richard      0          816       1  male   
822         Reuchlin, Jonkheer. John George      0          823       1  male   
1157  Chisholm, Mr. Roderick Robert Crispin      0         1158       1  male   
1263                Ismay, Mr. Joseph Bruce      0         1264       1  male   

      SibSp  Survived  Ticket  
179       0       0.0    LINE  
263       0       0.0  112059  
271       0       1.0    LINE  
277       0       0.0  239853  
302       0       0.0    LINE  
413       0       0.0  239853  
466       0       0.0  239853  
481       0       0.0  239854  
597       0       0.0    LINE  
633       0       0.0  112052  
674       0       0.0  239856  
732       0       0.0  239855  
806       0       0.0  112050  
815       0       0.0  112058  
822       0       0.0   19972  
1157      0       NaN  112051  
1263      0       NaN  112058

让我们先看下那些票价不为0的数据，其不同仓位等级的票均价

In [10]:
full.loc[full['Fare']!=0,:].groupby('Pclass')['Fare'].mean()

Pclass
1    89.447482
2    21.648108
3    13.378473
Name: Fare, dtype: float64

我们可以用这三个均值分别填充不同仓位其票价为0的记录,并用所有记录的均值填充na

In [11]:
full.loc[(full['Fare']==0)&(full['Pclass']==1),'Fare']=89.4
full.loc[(full['Fare']==0)&(full['Pclass']==2),'Fare']=21.6
full.loc[(full['Fare']==0)&(full['Pclass']==3),'Fare']=13.4
full['Fare']=full['Fare'].fillna(full['Fare'].mean())
full.describe()

Age         Fare        Parch  PassengerId       Pclass  \
count  1309.000000  1309.000000  1309.000000  1309.000000  1309.000000   
mean     29.881138    33.913981     0.385027   655.000000     2.294882   
std      12.883193    51.776484     0.865560   378.020061     0.837836   
min       0.170000     3.170800     0.000000     1.000000     1.000000   
25%      22.000000     7.925000     0.000000   328.000000     2.000000   
50%      29.881138    14.500000     0.000000   655.000000     3.000000   
75%      35.000000    31.387500     0.000000   982.000000     3.000000   
max      80.000000   512.329200     9.000000  1309.000000     3.000000   

             SibSp    Survived  
count  1309.000000  891.000000  
mean      0.498854    0.383838  
std       1.041658    0.486592  
min       0.000000    0.000000  
25%       0.000000    0.000000  
50%       0.000000    0.000000  
75%       1.000000    1.000000  
max       8.000000    1.000000

如果是分类数据，使用最常见的类别取代

In [12]:
#查看Embarked列中各value的数目
full['Embarked'].value_counts()

S    914
C    270
Q    123
Name: Embarked, dtype: int64

可以看到登船港口Embarked最常见的类别是"S"，故，使用其填充缺失项。

In [13]:
full['Embarked']=full['Embarked'].fillna('S')

如果是字符串类型，按照实际情况填写，无法追踪的信息，用"Unknow"填充。处理Cabin缺失值 U代表Unknow

In [14]:
full['Cabin']=full['Cabin'].fillna('U')

In [15]:
full.describe()

Age         Fare        Parch  PassengerId       Pclass  \
count  1309.000000  1309.000000  1309.000000  1309.000000  1309.000000   
mean     29.881138    33.913981     0.385027   655.000000     2.294882   
std      12.883193    51.776484     0.865560   378.020061     0.837836   
min       0.170000     3.170800     0.000000     1.000000     1.000000   
25%      22.000000     7.925000     0.000000   328.000000     2.000000   
50%      29.881138    14.500000     0.000000   655.000000     3.000000   
75%      35.000000    31.387500     0.000000   982.000000     3.000000   
max      80.000000   512.329200     9.000000  1309.000000     3.000000   

             SibSp    Survived  
count  1309.000000  891.000000  
mean      0.498854    0.383838  
std       1.041658    0.486592  
min       0.000000    0.000000  
25%       0.000000    0.000000  
50%       0.000000    0.000000  
75%       1.000000    1.000000  
max       8.000000    1.000000

In [16]:
full.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1309 entries, 0 to 1308
Data columns (total 12 columns):
Age            1309 non-null float64
Cabin          1309 non-null object
Embarked       1309 non-null object
Fare           1309 non-null float64
Name           1309 non-null object
Parch          1309 non-null int64
PassengerId    1309 non-null int64
Pclass         1309 non-null int64
Sex            1309 non-null object
SibSp          1309 non-null int64
Survived       891 non-null float64
Ticket         1309 non-null object
dtypes: float64(3), int64(4), object(5)
memory usage: 122.8+ KB


## 特征提取
如何知道哪些特征比较重要呢？通常需要与熟悉业务逻辑的人进行沟通，将业务人员说的特征反映到代码中，并通过实验和经验不断尝试，产生新的特征。

### Sex（性别）：

In [17]:
#将性别的值映射为数值
#male对应数值1，female对应数值0
sex_dict={'male':1,'female':0}
full['Sex']=full['Sex'].map(sex_dict)
full['Sex'].head()

0    1
1    0
2    0
3    0
4    1
Name: Sex, dtype: int64

### Embarked（登船港口）：
使用get_dummies进行one-hot编码

In [18]:
#使用get_dummies进行one-hot编码，产生虚拟变量（dummy variables），列名前缀(prefix)是Embarked
EmbarkedDf=pd.get_dummies(full['Embarked'],prefix='Embarked')
EmbarkedDf.head()

Embarked_C  Embarked_Q  Embarked_S
0           0           0           1
1           1           0           0
2           0           0           1
3           0           0           1
4           0           0           1

In [19]:
# 将EmbarkedDf的特征添加至full数据集
full=pd.concat([full,EmbarkedDf],axis=1)#axis=1表示按列插入数据
full.head()

Age Cabin Embarked     Fare  \
0  22.0     U        S   7.2500   
1  38.0   C85        C  71.2833   
2  26.0     U        S   7.9250   
3  35.0  C123        S  53.1000   
4  35.0     U        S   8.0500   

                                                Name  Parch  PassengerId  \
0                            Braund, Mr. Owen Harris      0            1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...      0            2   
2                             Heikkinen, Miss. Laina      0            3   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)      0            4   
4                           Allen, Mr. William Henry      0            5   

   Pclass  Sex  SibSp  Survived            Ticket  Embarked_C  Embarked_Q  \
0       3    1      1       0.0         A/5 21171           0           0   
1       1    0      1       1.0          PC 17599           1           0   
2       3    0      0       1.0  STON/O2. 3101282           0           0   
3       1    0      1       1.0            113803           0           0   
4       3    1      0       0.0            373450           0           0   

   Embarked_S  
0           1  
1           0  
2           1  
3           1  
4           1

因为已经使用登船港口(Embarked)进行了one-hot编码产生了它的虚拟变量（dummy variables）,
所以这里把登船港口(Embarked)删掉

In [20]:
full=full.drop('Embarked',axis=1)
full.head()

Age Cabin     Fare                                               Name  \
0  22.0     U   7.2500                            Braund, Mr. Owen Harris   
1  38.0   C85  71.2833  Cumings, Mrs. John Bradley (Florence Briggs Th...   
2  26.0     U   7.9250                             Heikkinen, Miss. Laina   
3  35.0  C123  53.1000       Futrelle, Mrs. Jacques Heath (Lily May Peel)   
4  35.0     U   8.0500                           Allen, Mr. William Henry   

   Parch  PassengerId  Pclass  Sex  SibSp  Survived            Ticket  \
0      0            1       3    1      1       0.0         A/5 21171   
1      0            2       1    0      1       1.0          PC 17599   
2      0            3       3    0      0       1.0  STON/O2. 3101282   
3      0            4       1    0      1       1.0            113803   
4      0            5       3    1      0       0.0            373450   

   Embarked_C  Embarked_Q  Embarked_S  
0           0           0           1  
1           1           0           0  
2           0           0           1  
3           0           0           1  
4           0           0           1

### Pclass(客舱等级)
方法同上

In [21]:
PclassDf=pd.get_dummies(full['Pclass'],prefix='Pclass')
PclassDf.head()

Pclass_1  Pclass_2  Pclass_3
0         0         0         1
1         1         0         0
2         0         0         1
3         1         0         0
4         0         0         1

In [22]:
full=pd.concat([full,PclassDf],axis=1)
full=full.drop('Pclass',axis=1)
full.head()

Age Cabin     Fare                                               Name  \
0  22.0     U   7.2500                            Braund, Mr. Owen Harris   
1  38.0   C85  71.2833  Cumings, Mrs. John Bradley (Florence Briggs Th...   
2  26.0     U   7.9250                             Heikkinen, Miss. Laina   
3  35.0  C123  53.1000       Futrelle, Mrs. Jacques Heath (Lily May Peel)   
4  35.0     U   8.0500                           Allen, Mr. William Henry   

   Parch  PassengerId  Sex  SibSp  Survived            Ticket  Embarked_C  \
0      0            1    1      1       0.0         A/5 21171           0   
1      0            2    0      1       1.0          PC 17599           1   
2      0            3    0      0       1.0  STON/O2. 3101282           0   
3      0            4    0      1       1.0            113803           0   
4      0            5    1      0       0.0            373450           0   

   Embarked_Q  Embarked_S  Pclass_1  Pclass_2  Pclass_3  
0           0           1         0         0         1  
1           0           0         1         0         0  
2           0           1         0         0         1  
3           0           1         1         0         0  
4           0           1         0         0         1

### Name(乘客姓名)：

In [23]:
full['Name'].head()

0                              Braund, Mr. Owen Harris
1    Cumings, Mrs. John Bradley (Florence Briggs Th...
2                               Heikkinen, Miss. Laina
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)
4                             Allen, Mr. William Henry
Name: Name, dtype: object

从上述Name字符串中发现每个名字里面都包含了头衔，我们可以获取到每个乘客的头衔，它可以帮助我们分析到更多有用的信息。

In [24]:
def getTitle(name):
    s1=name.split(',')[1]
    s2=s1.split('.')[0]
    return s2.strip()#移除字符串头尾空格
full['Title']=full['Name'].map(getTitle)
full['Title'].value_counts()

Mr              757
Miss            260
Mrs             197
Master           61
Rev               8
Dr                8
Col               4
Ms                2
Major             2
Mlle              2
Capt              1
Don               1
Mme               1
Jonkheer          1
Dona              1
Sir               1
the Countess      1
Lady              1
Name: Title, dtype: int64

将上述头衔对应到下面的几种类别中:

Officer政府官员；
Royalty王室（皇室）；
Mr已婚男士；
Mrs已婚妇女；
Miss年轻未婚女子；
Master有技能的人/教师

In [25]:
title_dict={"Capt":"Officer","Col":"Officer","Major":"Officer","Jonkheer":"Royalty","Don":"Royalty","Sir":"Royalty","Dr":"Officer","Rev":"Officer"
            ,"the Countess":"Royalty","Dona":"Royalty","Mme":"Mrs","Mlle":"Miss","Ms":"Mrs","Mr" :"Mr","Mrs" :"Mrs","Miss" :"Miss"
            ,"Master" :"Master", "Lady" : "Royalty"}
full['Title']=full['Title'].map(title_dict)
full.head()

Age Cabin     Fare                                               Name  \
0  22.0     U   7.2500                            Braund, Mr. Owen Harris   
1  38.0   C85  71.2833  Cumings, Mrs. John Bradley (Florence Briggs Th...   
2  26.0     U   7.9250                             Heikkinen, Miss. Laina   
3  35.0  C123  53.1000       Futrelle, Mrs. Jacques Heath (Lily May Peel)   
4  35.0     U   8.0500                           Allen, Mr. William Henry   

   Parch  PassengerId  Sex  SibSp  Survived            Ticket  Embarked_C  \
0      0            1    1      1       0.0         A/5 21171           0   
1      0            2    0      1       1.0          PC 17599           1   
2      0            3    0      0       1.0  STON/O2. 3101282           0   
3      0            4    0      1       1.0            113803           0   
4      0            5    1      0       0.0            373450           0   

   Embarked_Q  Embarked_S  Pclass_1  Pclass_2  Pclass_3 Title  
0           0           1         0         0         1    Mr  
1           0           0         1         0         0   Mrs  
2           0           1         0         0         1  Miss  
3           0           1         1         0         0   Mrs  
4           0           1         0         0         1    Mr

In [26]:
full['Title'].value_counts()

Mr         757
Miss       262
Mrs        200
Master      61
Officer     23
Royalty      6
Name: Title, dtype: int64

利用上述头衔数据框进行One-hot编码

In [27]:
TitleDf=pd.get_dummies(full['Title'])#One-hot编码
full=pd.concat([full,TitleDf],axis=1)#将特征添加至源数据集
full.head()

Age Cabin     Fare                                               Name  \
0  22.0     U   7.2500                            Braund, Mr. Owen Harris   
1  38.0   C85  71.2833  Cumings, Mrs. John Bradley (Florence Briggs Th...   
2  26.0     U   7.9250                             Heikkinen, Miss. Laina   
3  35.0  C123  53.1000       Futrelle, Mrs. Jacques Heath (Lily May Peel)   
4  35.0     U   8.0500                           Allen, Mr. William Henry   

   Parch  PassengerId  Sex  SibSp  Survived            Ticket   ...     \
0      0            1    1      1       0.0         A/5 21171   ...      
1      0            2    0      1       1.0          PC 17599   ...      
2      0            3    0      0       1.0  STON/O2. 3101282   ...      
3      0            4    0      1       1.0            113803   ...      
4      0            5    1      0       0.0            373450   ...      

   Pclass_1  Pclass_2  Pclass_3  Title  Master  Miss Mr  Mrs  Officer  Royalty  
0         0         0         1     Mr       0     0  1    0        0        0  
1         1         0         0    Mrs       0     0  0    1        0        0  
2         0         0         1   Miss       0     1  0    0        0        0  
3         1         0         0    Mrs       0     0  0    1        0        0  
4         0         0         1     Mr       0     0  1    0        0        0  

[5 rows x 23 columns]

In [28]:
full=full.drop(['Name','Title'],axis=1)#删掉不需要的列
full.head()

Age Cabin     Fare  Parch  PassengerId  Sex  SibSp  Survived  \
0  22.0     U   7.2500      0            1    1      1       0.0   
1  38.0   C85  71.2833      0            2    0      1       1.0   
2  26.0     U   7.9250      0            3    0      0       1.0   
3  35.0  C123  53.1000      0            4    0      1       1.0   
4  35.0     U   8.0500      0            5    1      0       0.0   

             Ticket  Embarked_C   ...     Embarked_S  Pclass_1  Pclass_2  \
0         A/5 21171           0   ...              1         0         0   
1          PC 17599           1   ...              0         1         0   
2  STON/O2. 3101282           0   ...              1         0         0   
3            113803           0   ...              1         1         0   
4            373450           0   ...              1         0         0   

   Pclass_3  Master  Miss  Mr  Mrs  Officer  Royalty  
0         1       0     0   1    0        0        0  
1         0       0     0   0    1        0        0  
2         1       0     1   0    0        0        0  
3         0       0     0   0    1        0        0  
4         1       0     0   1    0        0        0  

[5 rows x 21 columns]

### Cabin(客舱号)：
客场号的类别值是首字母，因此我们提取客舱号的首字母为特征。

In [29]:
full['Cabin']=full['Cabin'].map(lambda x:x[0])
full['Cabin'].value_counts()

U    1014
C      94
B      65
D      46
E      41
A      22
F      21
G       5
T       1
Name: Cabin, dtype: int64

In [30]:
CabinDf=pd.get_dummies(full['Cabin'],prefix='Cabin')
full=pd.concat([full,CabinDf],axis=1)
full=full.drop('Cabin',axis=1)
full.head()

Age     Fare  Parch  PassengerId  Sex  SibSp  Survived            Ticket  \
0  22.0   7.2500      0            1    1      1       0.0         A/5 21171   
1  38.0  71.2833      0            2    0      1       1.0          PC 17599   
2  26.0   7.9250      0            3    0      0       1.0  STON/O2. 3101282   
3  35.0  53.1000      0            4    0      1       1.0            113803   
4  35.0   8.0500      0            5    1      0       0.0            373450   

   Embarked_C  Embarked_Q   ...     Royalty  Cabin_A  Cabin_B  Cabin_C  \
0           0           0   ...           0        0        0        0   
1           1           0   ...           0        0        0        1   
2           0           0   ...           0        0        0        0   
3           0           0   ...           0        0        0        1   
4           0           0   ...           0        0        0        0   

   Cabin_D  Cabin_E  Cabin_F  Cabin_G  Cabin_T  Cabin_U  
0        0        0        0        0        0        1  
1        0        0        0        0        0        0  
2        0        0        0        0        0        1  
3        0        0        0        0        0        0  
4        0        0        0        0        0        1  

[5 rows x 29 columns]

### 建立家庭人数和家庭类别：
家庭人数=同代直系亲属数（Parch）+不同代直系亲属数（SibSp）+乘客自己（因为乘客自己也是家庭成员的一个，所以这里加1）

小家庭Family_Single： 家庭人数=1

中等家庭Family_Small: 2<=家庭人数<=4

大家庭Family_Large: 家庭人数>=5

In [31]:
full['familysize']=full['Parch']+full['SibSp']+1
full['family_singel']=np.where(full['familysize']==1,1,0)
full['family_small']=np.where((full['familysize']>=2)&(full['familysize']<=4),1,0)
full['family_large']=np.where(full['familysize']>=5,1,0)
full.head()

Age     Fare  Parch  PassengerId  Sex  SibSp  Survived            Ticket  \
0  22.0   7.2500      0            1    1      1       0.0         A/5 21171   
1  38.0  71.2833      0            2    0      1       1.0          PC 17599   
2  26.0   7.9250      0            3    0      0       1.0  STON/O2. 3101282   
3  35.0  53.1000      0            4    0      1       1.0            113803   
4  35.0   8.0500      0            5    1      0       0.0            373450   

   Embarked_C  Embarked_Q      ...       Cabin_D  Cabin_E  Cabin_F  Cabin_G  \
0           0           0      ...             0        0        0        0   
1           1           0      ...             0        0        0        0   
2           0           0      ...             0        0        0        0   
3           0           0      ...             0        0        0        0   
4           0           0      ...             0        0        0        0   

   Cabin_T  Cabin_U  familysize  family_singel  family_small  family_large  
0        0        1           2              0             1             0  
1        0        0           2              0             1             0  
2        0        1           1              1             0             0  
3        0        0           2              0             1             0  
4        0        1           1              1             0             0  

[5 rows x 33 columns]

In [32]:
full.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1309 entries, 0 to 1308
Data columns (total 33 columns):
Age              1309 non-null float64
Fare             1309 non-null float64
Parch            1309 non-null int64
PassengerId      1309 non-null int64
Sex              1309 non-null int64
SibSp            1309 non-null int64
Survived         891 non-null float64
Ticket           1309 non-null object
Embarked_C       1309 non-null uint8
Embarked_Q       1309 non-null uint8
Embarked_S       1309 non-null uint8
Pclass_1         1309 non-null uint8
Pclass_2         1309 non-null uint8
Pclass_3         1309 non-null uint8
Master           1309 non-null uint8
Miss             1309 non-null uint8
Mr               1309 non-null uint8
Mrs              1309 non-null uint8
Officer          1309 non-null uint8
Royalty          1309 non-null uint8
Cabin_A          1309 non-null uint8
Cabin_B          1309 non-null uint8
Cabin_C          1309 non-null uint8
Cabin_D          1309 non-null uint8

In [33]:
full.loc[889:898,:]

Age     Fare  Parch  PassengerId  Sex  SibSp  Survived   Ticket  \
889  26.0  30.0000      0          890    1      0       1.0   111369   
890  32.0   7.7500      0          891    1      0       0.0   370376   
891  34.5   7.8292      0          892    1      0       NaN   330911   
892  47.0   7.0000      0          893    0      1       NaN   363272   
893  62.0   9.6875      0          894    1      0       NaN   240276   
894  27.0   8.6625      0          895    1      0       NaN   315154   
895  22.0  12.2875      1          896    0      1       NaN  3101298   
896  14.0   9.2250      0          897    1      0       NaN     7538   
897  30.0   7.6292      0          898    0      0       NaN   330972   
898  26.0  29.0000      1          899    1      1       NaN   248738   

     Embarked_C  Embarked_Q      ...       Cabin_D  Cabin_E  Cabin_F  Cabin_G  \
889           1           0      ...             0        0        0        0   
890           0           1      ...             0        0        0        0   
891           0           1      ...             0        0        0        0   
892           0           0      ...             0        0        0        0   
893           0           1      ...             0        0        0        0   
894           0           0      ...             0        0        0        0   
895           0           0      ...             0        0        0        0   
896           0           0      ...             0        0        0        0   
897           0           1      ...             0        0        0        0   
898           0           0      ...             0        0        0        0   

     Cabin_T  Cabin_U  familysize  family_singel  family_small  family_large  
889        0        0           1              1             0             0  
890        0        1           1              1             0             0  
891        0        1           1              1             0             0  
892        0        1           2              0             1             0  
893        0        1           1              1             0             0  
894        0        1           1              1             0             0  
895        0        1           3              0             1             0  
896        0        1           1              1             0             0  
897        0        1           1              1             0             0  
898        0        1           3              0             1             0  

[10 rows x 33 columns]

## 特征选择和特征降维
通过前面的特征选取，得到32个特征，下面使用相关系数法选取特征

In [37]:
#计算相关性矩阵
corr_df=full.corr()
corr_df

Age      Fare     Parch  PassengerId       Sex     SibSp  \
Age            1.000000  0.176326 -0.130872     0.025731  0.057397 -0.190747   
Fare           0.176326  1.000000  0.216043     0.034120 -0.176464  0.154383   
Parch         -0.130872  0.216043  1.000000     0.008942 -0.213125  0.373587   
PassengerId    0.025731  0.034120  0.008942     1.000000  0.013406 -0.055224   
Sex            0.057397 -0.176464 -0.213125     0.013406  1.000000 -0.109609   
SibSp         -0.190747  0.154383  0.373587    -0.055224 -0.109609  1.000000   
Survived      -0.070323  0.246552  0.081629    -0.005007 -0.543351 -0.035322   
Embarked_C     0.076179  0.279941 -0.008635     0.048101 -0.066564 -0.048396   
Embarked_Q    -0.012718 -0.133808 -0.100943     0.011585 -0.088651 -0.048678   
Embarked_S    -0.059153 -0.161943  0.071881    -0.049836  0.115193  0.073709   
Pclass_1       0.362587  0.614106 -0.013033     0.026495 -0.107371 -0.034256   
Pclass_2      -0.014193 -0.122792 -0.010057     0.022714 -0.028862 -0.052419   
Pclass_3      -0.302093 -0.430696  0.019521    -0.041544  0.116562  0.072610   
Master        -0.363923  0.008946  0.253482     0.002254  0.164375  0.329171   
Miss          -0.254146  0.086006  0.066473    -0.050027 -0.672819  0.077564   
Mr             0.165476 -0.184520 -0.304780     0.014116  0.870678 -0.243104   
Mrs            0.198091  0.134059  0.213491     0.033299 -0.571176  0.061643   
Officer        0.162818  0.027077 -0.032631     0.002231  0.087288 -0.013813   
Royalty        0.059466  0.044920 -0.030197     0.004400 -0.020408 -0.010787   
Cabin_A        0.125177  0.028783 -0.030707    -0.002831  0.047561 -0.039808   
Cabin_B        0.113458  0.408948  0.073051     0.015895 -0.094453 -0.011569   
Cabin_C        0.167993  0.397754  0.009601     0.006092 -0.077473  0.048616   
Cabin_D        0.132886  0.070403 -0.027385     0.000549 -0.057396 -0.015727   
Cabin_E        0.106600  0.071746  0.001084    -0.008136 -0.040340 -0.027180   
Cabin_F       -0.072644 -0.039065  0.020481     0.000306 -0.006655 -0.008619   
Cabin_G       -0.085977 -0.023580  0.058325    -0.045949 -0.083285  0.006015   
Cabin_T        0.032461  0.000847 -0.012304    -0.023049  0.020558 -0.013247   
Cabin_U       -0.271918 -0.513016 -0.036806     0.000208  0.137396  0.009064   
familysize    -0.196996  0.219629  0.792296    -0.031437 -0.188583  0.861952   
family_singel  0.116675 -0.264940 -0.549022     0.028546  0.284537 -0.591077   
family_small  -0.038189  0.188678  0.248532     0.002975 -0.255196  0.253590   
family_large  -0.161210  0.167639  0.624627    -0.063415 -0.077748  0.699681   

               Survived  Embarked_C  Embarked_Q  Embarked_S      ...       \
Age           -0.070323    0.076179   -0.012718   -0.059153      ...        
Fare           0.246552    0.279941   -0.133808   -0.161943      ...        
Parch          0.081629   -0.008635   -0.100943    0.071881      ...        
PassengerId   -0.005007    0.048101    0.011585   -0.049836      ...        
Sex           -0.543351   -0.066564   -0.088651    0.115193      ...        
SibSp         -0.035322   -0.048396   -0.048678    0.073709      ...        
Survived       1.000000    0.168240    0.003650   -0.149683      ...        
Embarked_C     0.168240    1.000000   -0.164166   -0.778262      ...        
Embarked_Q     0.003650   -0.164166    1.000000   -0.491656      ...        
Embarked_S    -0.149683   -0.778262   -0.491656    1.000000      ...        
Pclass_1       0.285904    0.325722   -0.166101   -0.181800      ...        
Pclass_2       0.093349   -0.134675   -0.121973    0.196532      ...        
Pclass_3      -0.322308   -0.171430    0.243706   -0.003805      ...        
Master         0.085221   -0.014172   -0.009091    0.018297      ...        
Miss           0.332795   -0.014351    0.198804   -0.113886      ...        
Mr            -0.549199   -0.065538   -0.080224    0.108924      ...        
Mrs            0.344935    0.098379   -0.100374   -0.022950      ...  

In [40]:
#提取各特征与生存情况（Survived）的相关系数，并降序排列
corr_df['Survived'].sort_values(ascending=False)

Survived         1.000000
Mrs              0.344935
Miss             0.332795
Pclass_1         0.285904
family_small     0.279855
Fare             0.246552
Cabin_B          0.175095
Embarked_C       0.168240
Cabin_D          0.150716
Cabin_E          0.145321
Cabin_C          0.114652
Pclass_2         0.093349
Master           0.085221
Parch            0.081629
Cabin_F          0.057935
Royalty          0.033391
Cabin_A          0.022287
familysize       0.016639
Cabin_G          0.016040
Embarked_Q       0.003650
PassengerId     -0.005007
Cabin_T         -0.026456
Officer         -0.031316
SibSp           -0.035322
Age             -0.070323
family_large    -0.125147
Embarked_S      -0.149683
family_singel   -0.203367
Cabin_U         -0.316912
Pclass_3        -0.322308
Sex             -0.543351
Mr              -0.549199
Name: Survived, dtype: float64

根据各特征与生存情况（Survived）的相关系数大小，选取以下特征进行建模：头衔（前面所在的数据集TitleDf）、客舱等级（PclassDf）、船票价格（Fare）、船舱号（CabinDf）、登船港口（EmbarkedDf）、性别（Sex）、家庭大小及类别（familysize,family_small,family_large,family_singel）

In [42]:
full_x=pd.concat([TitleDf,PclassDf,CabinDf,EmbarkedDf,full['Fare'],full['Sex'],full['familysize'],full['family_small']
                  ,full['family_large'],full['family_singel']],axis=1)
full_x.head()

Master  Miss  Mr  Mrs  Officer  Royalty  Pclass_1  Pclass_2  Pclass_3  \
0       0     0   1    0        0        0         0         0         1   
1       0     0   0    1        0        0         1         0         0   
2       0     1   0    0        0        0         0         0         1   
3       0     0   0    1        0        0         1         0         0   
4       0     0   1    0        0        0         0         0         1   

   Cabin_A      ...        Cabin_U  Embarked_C  Embarked_Q  Embarked_S  \
0        0      ...              1           0           0           1   
1        0      ...              0           1           0           0   
2        0      ...              1           0           0           1   
3        0      ...              0           0           0           1   
4        0      ...              1           0           0           1   

      Fare  Sex  familysize  family_small  family_large  family_singel  
0   7.2500    1           2             1             0              0  
1  71.2833    0           2             1             0              0  
2   7.9250    0           1             0             0              1  
3  53.1000    0           2             1             0              0  
4   8.0500    1           1             0             0              1  

[5 rows x 27 columns]

## 构建模型
### 建立训练数据集和测试数据集
根据前面的数据我们知道，train.csv里包含Survived标签，因此用来作为模型训练的数据，并需要将其分为训练数据集和测试数据集，test.csv无Survived标签，用来作为预测数据集

In [44]:
#前891行为原始训练数据，我们将其提取出来
source_x=full_x.loc[0:890,:]#提取特征
source_y=full.loc[0:890,'Survived']#提取标签
#后418行为预测数据
pred_x=full_x.loc[891:,:]
source_x.shape
source_y.shape
pred_x.shape

(891, 27)

(891,)

(418, 27)

In [67]:
#建立模型用的训练数据集和测试数据集，按照二八原则分为训练数据和测试数据，其中80%为训练数据
from sklearn.cross_validation import train_test_split
train_x,test_x,train_y,test_y=train_test_split(source_x,source_y,train_size=0.8)
print('训练数据集特征:{0},训练数据集标签:{1}'.format(train_x.shape,train_y.shape))
print('测试数据集特征:{0},测试数据集标签:{1}'.format(test_x.shape,test_y.shape))

训练数据集特征:(712, 27),训练数据集标签:(712,)
测试数据集特征:(179, 27),测试数据集标签:(179,)


In [217]:
#对train_x,test_x进行标准化
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
train_x_std=sc.fit_transform(train_x)
test_x_std=sc.transform(test_x)

### 选择算法训练模型
这里我们选择逻辑回归

In [218]:
#第一步：选择算法，并导入相应算发包
from sklearn.linear_model import LogisticRegression
#第二步：创建模型
model=LogisticRegression()
#第三步：训练模型
model.fit(train_x_std,train_y)

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='ovr', n_jobs=1,
          penalty='l2', random_state=None, solver='liblinear', tol=0.0001,
          verbose=0, warm_start=False)

### 评估模型

In [219]:
#得出模型正确率
model.score(test_x_std,test_y)

0.8212290502793296

### 方案实施

In [220]:
#使用训练得到的模型对pred_x的生存情况进行预测
pred_x_std=sc.fit_transform(pred_x)
pred_y=model.predict(pred_x_std)
pred_y

array([0., 1., 0., 0., 1., 0., 1., 0., 1., 0., 0., 0., 1., 0., 1., 1., 0.,
       0., 1., 1., 0., 1., 1., 0., 1., 0., 1., 0., 0., 0., 0., 0., 1., 1.,
       0., 0., 1., 1., 0., 0., 0., 0., 0., 1., 1., 0., 0., 0., 1., 1., 0.,
       0., 1., 0., 0., 0., 0., 0., 0., 1., 0., 0., 0., 1., 1., 1., 1., 0.,
       0., 1., 1., 0., 1., 1., 1., 0., 0., 1., 0., 1., 1., 0., 0., 0., 0.,
       0., 1., 1., 1., 1., 1., 0., 1., 0., 0., 0., 1., 0., 1., 0., 1., 0.,
       0., 0., 1., 0., 0., 0., 0., 0., 0., 1., 1., 1., 1., 0., 0., 0., 0.,
       1., 1., 0., 1., 0., 0., 1., 0., 1., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 1., 0., 0., 1., 0., 0., 0., 0., 1., 0., 0., 0., 1., 0., 0.,
       1., 0., 0., 1., 1., 0., 1., 1., 1., 1., 0., 0., 1., 0., 0., 1., 1.,
       0., 0., 0., 0., 0., 1., 1., 0., 1., 1., 0., 1., 1., 0., 1., 0., 1.,
       0., 0., 0., 0., 0., 1., 0., 1., 0., 1., 1., 0., 1., 1., 1., 1., 1.,
       0., 0., 1., 0., 1., 0., 0., 0., 0., 1., 0., 0., 1., 0., 1., 0., 1.,
       0., 1., 0., 1., 1.

In [221]:
pred_df=pd.DataFrame({'PassengerId':test.PassengerId,'Survived':pred_y})
pred_df.head()

PassengerId  Survived
0          892       0.0
1          893       1.0
2          894       0.0
3          895       0.0
4          896       1.0

In [222]:
pred_df.shape

(418, 2)

In [223]:
pred_df['Survived']=pred_df['Survived'].astype('int')
pred_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 418 entries, 0 to 417
Data columns (total 2 columns):
PassengerId    418 non-null int64
Survived       418 non-null int32
dtypes: int32(1), int64(1)
memory usage: 5.0 KB


In [224]:
#保存结果
pred_df.to_csv(r'E:\python\data\titanic\predict.csv',index=False)